Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm

```diff
@@ -3192,6 +3192,124 @@ - (void)_updateFixedColorExtensionViewFrames
+                    completionHandler(nullptr, adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:userInfo]).get());
+                    return;
+                }
+                encoder << data->sessionStorage.value();
+                break;
+            default:
+                ASSERT_NOT_REACHED();
+                break;
+            }
+        }
+
+        completionHandler(toNSData(encoder.span()).get(), nullptr);
+    });
+
+    if (dataTypes & _WKWebViewDataTypeSessionStorage) {
+        RefPtr page = [self _protectedPage];
+        page->fetchSessionStorage([callbackAggregator, protectedPage = page, data](auto&& sessionStorage) {
+            data->sessionStorage = WTFMove(sessionStorage);
+        });
+    }
+}
+
+- (void)restoreData:(NSData *)data completionHandler:(void(^)(NSError *))completionHandler
+{
+    WTF::Persistence::Decoder decoder(span(data));
+
+    std::optional<unsigned> currentWKWebViewDataSerializationVersion;
+    decoder >> currentWKWebViewDataSerializationVersion;
+    if (!currentWKWebViewDataSerializationVersion) {
+        NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : @"Data is in incorrect format.", };
+        completionHandler(adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:userInfo]).get());
+        return;
+    }
+
+    std::optional<Vector<WebKit::WebViewDataType>> encodedDataTypes;
+    decoder >> encodedDataTypes;
+    if (!encodedDataTypes) {
+        NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : @"Data is in incorrect format.", };
+        completionHandler(adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:userInfo]).get());
+        return;
+    }
+
+    auto succeeded = Box<bool>::create(true);
+    Ref callbackAggregator = CallbackAggregator::create([completionHandler = makeBlockPtr(completionHandler), succeeded] {
+        if (*succeeded)
+            completionHandler(nullptr);
+        else {
+            NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : @"Unknown error occurred while restoring data.", };
+            completionHandler(adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:userInfo]).get());
+        }
+    });
+
+    for (auto& encodedDataType : *encodedDataTypes) {
+        switch (encodedDataType) {
+        case WebKit::WebViewDataType::SessionStorage: {
+            std::optional<HashMap<WebCore::ClientOrigin, HashMap<String, String>>> sessionStorage;
+            decoder >> sessionStorage;
+
+            if (!sessionStorage) {
+                NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : @"Data is in incorrect format.", };
+                completionHandler(adoptNS([[NSError alloc] initWithDomain:WKErrorDomain code:NSCoderReadCorruptError userInfo:userInfo]).get());
```

Comment:
Does not seem right to invoke `completionHandler` after it's moved to `callbackAggregator`.
You might construct `WKWebViewData` from the encoded data before creating `CallbackAggregator` with `CompletionHandler` and invoking `restoreSessionStorage`.

Same issue for the API in `WKWebsiteDataStore`.

Reply:
Fixed!